text/0000-xor-urls/0000-xor-urls.md

```diff
@@ -0,0 +1,115 @@
+```
+<safe-url>            := 'safe://' ( <xor-uri> | <public-name-uri> )
+
+<public-name-uri>     :=  [<service> '.'] <public-name> <path-query-fragment>
+
+<xor-uri>             := <immutable-data-uri> | <mutable-data-uri>
+
+<immutable-data-uri>  := <cid> <query-fragment>
+
+<mutable-data-uri>    := <cid> ':' <type-tag> ['+' <content-version>] <path-query-fragment>
+
+<path-query-fragment> := ['/' <path>] <query-fragment>
+
+<query-fragment>      := ['?' <query>] ['#' <fragment>]
+```
+
+Where:
+
+- `<cid>`: follows the [CID format](https://github.com/ipld/cid) which self-describes and encodes:
+    - the base encoding for the string, it's here proposed to use [z-base32 encoding](http://philzimmermann.com/docs/human-oriented-base-32-encoding.txt) for the reasons explained above,
+    - the version of CID for upgradability purposes (v1 now),
+    - the content type or codec of the data being referenced,
+    - the XOR address of the content being referenced
+- `<content-version>`: for future implementation to reference versionable content, using a single address with different versions
+
+- `<type-tag>`: the type tag value if the CID is referencing a MutableData. In the absence of this value, the CID will be assumed to be targetting an ImmutableData
+
+- `<path>`: the path of the file if the CID is referencing a MutableData, which can be accessed through the NFS/Files convention (or other emulations/conventions in the future) and resolve it
+
+- `<query>`: query arguments, to be used by the client app and not for retrieving the content
+
+- `<fragment>`: fragment of the content, to be used by the client app and not for retrieving the content
+
+The resolver function shall first attempt to decode the URL assuming it contains a `<cid>` part. If that step fails (either because it couldn’t decode the CID, or because there was no data at the decoded XOR address), it will do a fallback to assume it’s a public-name URL and it shall try to find the pulic-name MD (with the `sha3` hash of the string), just like any public-name URLs are resolved now by the resolver function.
+
+The following are examples of what would become valid XOR-URLs as per this proposal:
+
+ImmutableData XOR-URL:
+`safe://hygjdkfty6m7ag3bckq7eqgeizbtjk915c3jbrcgtisad8iikbk4xws4jbpky`
+
+MutableData XOR-URL:
+`safe://hyfktce8j75yhmj1dbi1xw5wnb4m3zdydr7wpbzf1a16hc3sbxzu8a9hiqw:15000`
+
+NFS/Files container MutableData XOR-URL:
+`safe://hyfktcenm57js4bm3owhez9td9pi3t8bzk1crqp7mr5865c15ih3yxpz68w:15008/some/folder/index.html`
+
+## Drawbacks
+
+None identified as of yet.
+
+## Alternatives
+
```

Comment:
Another alternative is a straight up `safe-xor://` protocol for IDing these URIs... Maybe good to mention why we dont want that and why CID as id of XOR is enough.

Reply:
oh yes, good point, I'll add that!